Add blosc getitem

andrewfulton9 opened the pull request:
This PR demonstrates a way to do partial decompression of a blosc compressed array buffer. The biggest limitation of the `decompress_partial` method is the inability to properly decompress parts of a multidimensional array. The user must also must account for typesize and the encoding size of the encoded array to function properly.

TODO:
* [x] Unit tests and/or doctests in docstrings
* [x] ``tox -e py38`` passes locally
* [x] Docstrings and API docs for any new/modified user-facing classes and functions
* [x] Changes documented in docs/release.rst
* [x] ``tox -e docs`` passes locally
* [ ] AppVeyor and Travis CI passes
* [ ] Test coverage to 100% (Coveralls passes)

andrewfulton9 committed:
adds partial decompression, not passing tests

andrewfulton9 committed:
work

andrewfulton9 committed:
debugging

andrewfulton9 committed:
passes tests

andrewfulton9 committed:
turns back default vebosity on tests

andrewfulton9 committed:
updates docs

Carreau commented on numcodecs/blosc.pyx:
```diff
@@ -393,6 +394,83 @@ def decompress(source, dest=None):
+    dest : array-like, optional
+        Object to decompress into.
+    start: int,
+        Offset in inem where we want to start decoding
```
```suggestion
        Offset in item where we want to start decoding
```

Carreau commented on numcodecs/blosc.pyx:
```diff
@@ -393,6 +394,83 @@ def decompress(source, dest=None):
+        
+        if dest_nbytes < nbitems_bytes:
+            raise ValueError('destination buffer too small; expected at least %s, '
+                             'got %s' % (nitems, dest_nbytes))
```
Those are likely the wrong variable now.

Carreau commented on numcodecs/blosc.pyx:
```diff
@@ -393,6 +394,83 @@ def decompress(source, dest=None):
+            dest_buffer.release()
+
+    if ret <= 0:
+        raise RuntimeError('error during blosc partial decompression: %d, start: %d, nbytes: %d' % (ret, start, nbytes))
```
May want to describe what `ret` is there.

Carreau commented on numcodecs/blosc.pyx:
```diff
@@ -486,6 +564,11 @@ class Blosc(Codec):
+        buf = ensure_contiguous_ndarray(buf, self.max_buffer_size)
+        return decompress_partial(buf, start, nitems,
+                typesize, encoding_size, dest=out)
+
```
Here we _could_ try to add an option to "guess" the encoding size, if we know the typesize then we know blosc will return _at least_ `typesize` bytes if we ask to decode the first element. 

IIRC when blosc_getitem succeed it tells you how much it wrote to the buffer you gave it, so if  you might be able to figure out whether the encoded typesize is correct, or 1 (if stored as bytes). 

Not sure it is worth it though and there is a performance impact to chec that as you need to decompress twice.

Carreau replied:
We could also just peek at buf[3]... 
https://github.com/Blosc/c-blosc/blob/403d50ced2cd35d3e72a007d29d73110ac45e62b/blosc/blosc.c#L1613

andrewfulton9 replied:
I added some code to infer the typesize / encoding size as you suggested here, and its passing the tests. I changed typesize and encoding size to kwargs that default to 0, and if they are zero, they use the inferences to do the partial decompression

Carreau commented on numcodecs/tests/common.py:
```diff
@@ -31,6 +31,8 @@ def compare_arrays(arr, res, precision=None):
         order = 'F'
     else:
         order = 'C'
+    print(arr.shape)
+    print(res.shape)
```
```suggestion
```

Carreau commented on numcodecs/tests/test_blosc.py:
```diff
@@ -58,6 +61,14 @@ def test_encode_decode():
 
 
+@pytest.mark.parametrize('codec', codecs)
+@pytest.mark.parametrize('array', [pytest.param(x) if len(x.shape) == 1
+                                   else pytest.param(x, marks=[pytest.mark.xfail])
+                                   for x in arrays])
```
Oh, nice  !

Carreau commented:
That looks good to me, I'm thinking that when this is used in a higher level API we might be able to guess the typesize of the underlying blosc serialization once and reuse it – this would then let the users be unaware of this detail.

andrewfulton9 committed:
adds code to infer typesize/encoding size, cleans up code

andrewfulton9 committed:
removes unused import

Carreau committed:
Example of unstable-allowing context manager.

The advantage of using warnings instead of globals is performance and
the ability to also still get the warning printed by default.

andrewfulton9 committed:
Merge pull request #2 from Carreau/warning-context

Example of unstable-allowing context manager.

andrewfulton9 committed:
removes unstable context manager

andrewfulton9 committed:
adds experimental docstrings

andrewfulton9 committed:
updates cython c generated file

andrewfulton9 committed:
removes extra line

andrewfulton9 committed:
rollback c file noise changes

andrewfulton9 committed:
encoding size is taken from source header now

andrewfulton9 committed:
fix method docstring

andrewfulton9 committed:
removes debugging print statements

andrewfulton9 committed:
removes encoding_size argument from partial_decompression, updates tests

andrewfulton9 committed:
fixes up formatting of docstrings. Adds check that destination buffer is the right size

andrewfulton9 committed:
merges

alimanfoo committed:
Update docs/release.rst
